splitter/splitter.go

```diff
@@ -1,39 +1,147 @@
+// Package splitter provides functions for decoding various kinds of records that might come off of a kinesis stream.
+// It is equipped to with the functions to unbundle KPL aggregates and CloudWatch log bundles,
+// as well as apply appropriate decompression.
+// KCL applications would be most interested in `SplitMessageIfNecessary` which can handle zlibbed records as well as
+// CloudWatch bundles. KCL automatically unbundles KPL aggregates before passing the records to the consumer.
+// Non-KCL applications (such as Lambdas consuming KPL-produced aggregates) should either use
+// - KPLDeaggregate if the consumer purely wants to unbundle KPL aggregates, but will handle the raw records themselves.
+// - DeaggregateAndSplitIfNecessary if the consumer wants to apply the same decompress and split logic as SplitMessageIfNecessary
+//   in addition to the KPL splitting.
 package splitter
 
 import (
 	"bytes"
 	"compress/gzip"
 	"compress/zlib"
+	"crypto/md5"
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"regexp"
 	"strconv"
 	"time"
+
+	kpl "github.com/a8m/kinesis-producer"
+	"github.com/golang/protobuf/proto"
 )
 
-// SplitMessageIfNecessary handles three types of records:
+// The Amazon Kinesis Producer Library (KPL) aggregates multiple logical user records into a single
+// Amazon Kinesis record for efficient puts.
+// https://github.com/awslabs/amazon-kinesis-producer/blob/master/aggregation-format.md
+var kplMagicNumber = []byte{0xF3, 0x89, 0x9A, 0xC2}
+
+// IsKPLAggregate checks a record for a KPL aggregate prefix.
+// It is not necessary to call this before calling KPLDeaggregate.
+func IsKPLAggregate(data []byte) bool {
+	return bytes.HasPrefix(data, kplMagicNumber)
+}
+
+// KPLDeaggregate takes a Kinesis record and converts it to one or more user records by applying KPL deaggregation.
+// If the record begins with the 4-byte magic prefix that KPL uses, the single Kinesis record is split into its component user records.
+// Otherwise, the return value is a singleton containing the original record.
+func KPLDeaggregate(kinesisRecord []byte) ([][]byte, error) {
+	if !IsKPLAggregate(kinesisRecord) {
+		return [][]byte{kinesisRecord}, nil
```

Comment:
I don't think I'm understanding why this version would ever get used currently. Would it make more sense to move this case into `SplitMessageIfNecessary` and then make `Deaggregate` look like the following?

```
if IsKPLAggregate(record) {
    return KPLDeaggregate(record)
}
return SplitMessageIfNecessary(record)
```

Reply:
Actually now that I look at it `unzlib` already seems to handle that case.

Reply:
We have no use for it right now, it's just a utility function that someone using this package might want (e.g. for non-log data on kinesis). It was free to include, so I figured why not.

Reply:
I think it's slightly confusing because I'm not sure I would know which of them to choose. And if the record isn't gzipped `Deaggregate` already yields the same output.

Now that I look closer, if the record IS gzipped, `SplitMessageIfNecessary` assumes it has a specific format. But `KPLDeaggregate` will just forward along the gzipped record which doesn't seem great either.

Reply:
Does the package comment at the top not explain it well enough?